File: frontend/src/types/project.ts
import type { Category, ID, ISODateString, PublishStatus, Technology } from "./common";

export interface Project {
  id: ID;
  title: string;
  slug: string;
  description: string;
  content?: string;
  thumbnail?: string;
  gallery?: string[];
  category: Category;
  technologies: Technology[];
  github_url?: string;
  live_url?: string;
  featured: boolean;
  status: PublishStatus;
  year: string;
  start_date?: ISODateString;
  end_date?: ISODateString;
  created_at: ISODateString;
  updated_at: ISODateString;
}

export type ProjectInput = Omit<Project, "id" | "created_at" | "updated_at" | "category" | "technologies"> & {
  category_id: ID;
  technology_ids: ID[];
};

File: backend/app/db/models/project.py
from datetime import datetime

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Table, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

project_technology = Table(
    "project_technology",
    Base.metadata,
    Column("project_id", Integer, ForeignKey("projects.id"), primary_key=True),
    Column("technology_id", Integer, ForeignKey("technologies.id"), primary_key=True),
)


class Project(Base):
    __tablename__ = "projects"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    slug: Mapped[str] = mapped_column(String(255), unique=True, index=True, nullable=False)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    project_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    github_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    featured: Mapped[bool] = mapped_column(default=False)
    status: Mapped[str] = mapped_column(String(50), default="published")
    order_index: Mapped[int] = mapped_column(default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    technologies: Mapped[list["Technology"]] = relationship(secondary=project_technology, back_populates="projects")


class Technology(Base):
    __tablename__ = "technologies"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(255), unique=True, nullable=False)
    category: Mapped[str | None] = mapped_column(String(100), nullable=True)
    projects: Mapped[list[Project]] = relationship(secondary=project_technology, back_populates="technologies")
